mixer/adapter/kubernetesenv/kubernetesenv.go

```diff
@@ -299,3 +351,52 @@ func newKubernetesClient(kubeconfigPath string, env adapter.Env) (k8s.Interface,
 	}
 	return k8s.NewForConfig(config)
 }
+
+func (b *builder) createCacheController(k8sInterface k8s.Interface, clusterID string) error {
+	controller, err := getNewCacheController(b, k8sInterface, b.kubeHandler.env)
+	if err == nil {
+		b.controllers[clusterID] = controller
+		b.kubeHandler.k8sCache[clusterID] = controller
+		b.kubeHandler.env.Logger().Infof("created remote controller %s", clusterID)
+	} else {
+		b.kubeHandler.env.Logger().Errorf("error on creating remote controller %s err = %v", clusterID, err)
+	}
+
+	return err
+}
+
+func (b *builder) deleteCacheController(clusterID string) error {
+	b.kubeHandler.k8sCache[clusterID].StopControlChannel()
+	delete(b.controllers, clusterID)
+	delete(b.kubeHandler.k8sCache, clusterID)
+
+	b.kubeHandler.env.Logger().Infof("deleted remote controller %s", clusterID)
+
+	return nil
+}
+
+func initMultiClusterSecretController(b *builder, kubeconfig string, env adapter.Env) (err error) {
+	var clusterNS string
+
+	paramsProto := b.adapterConfig
+	if paramsProto.ClusterRegistriesNamespace == "" {
+		clusterNS = os.Getenv("POD_NAMESPACE")
+		if clusterNS == "" {
+			clusterNS = clusterRegistriesNamespace
+		}
+	} else {
+		clusterNS = paramsProto.ClusterRegistriesNamespace
+	}
+
+	kubeClient, err := b.newClientFn(kubeconfig, env)
+	if err != nil {
+		return fmt.Errorf("could not create K8s client: %v", err)
+	}
+
+	err = secretcontroller.StartSecretController(kubeClient, b.createCacheController, b.deleteCacheController, clusterNS)
```

Comment:
`StartSecretController` creates a goroutine, correct? @geeknoid, what is the guidance on calls that create child goroutines for adapters these days?

Reply:
Yes. the multi-cluster secret controller code originally written just for pilot has been made common (pkg/kube/secretcontroller/secretcontroller.go) so that it can be used by multiple istio sources (this is the PR for mixer usage, https://github.com/istio/istio/pull/8479 is in flight to update pilot to use this common package). The multi-cluster secret controller waits for secret updates made to secrets created w/ the label 'istio/multiCluster=true', updates a few local data structures then uses callbacks in inform the consumer of the event.

Reply:
For in-process adapters, please use Env.ScheduleWork or Env.ScheduleDaemon. Of course, this only works for top-level goroutines. If you are using libraries that spawn their own goroutines, that's OK too. But if they crash, all of Mixer will go down, as opposed to goroutines started with ScehduleDaemon and ScheduleWork will contain their blast radius.

Reply:
OK, I'll investigate and make changes under a different PR. Like to keep this one to only the required changes for mixer k8s multi-cluster support.